pkg/operator/controllers/monitoring/monitoring_controller.go

```diff
@@ -0,0 +1,212 @@
+		}
+
+		changed := false
+		if configData.PrometheusK8s.Retention != "15d" {
+			configData.PrometheusK8s.Retention = "15d"
+			changed = true
+		}
+
+		if configData.PrometheusK8s.VolumeClaimTemplate.Spec.Resources.Requests.Storage != "100Gi" {
+			configData.PrometheusK8s.VolumeClaimTemplate.Spec.Resources.Requests.Storage = "100Gi"
+			changed = true
+		}
+
+		if !isCreate && !changed {
+			return nil
+		}
+
+		var b []byte
+		err = codec.NewEncoderBytes(&b, &codec.JsonHandle{}).Encode(configData)
+		if err != nil {
+			return err
+		}
+
+		cmYaml, err := yaml.JSONToYAML(b)
+		if err != nil {
+			return err
+		}
+		cm.Data["config.yaml"] = string(cmYaml)
+
+		if isCreate {
+			r.log.Info("re-creating monitoring configmap")
+			_, err = r.kubernetescli.CoreV1().ConfigMaps(monitoringName.Namespace).Create(ctx, cm, metav1.CreateOptions{})
+		} else {
+			r.log.Info("updating monitoring configmap")
+			_, err = r.kubernetescli.CoreV1().ConfigMaps(monitoringName.Namespace).Update(ctx, cm, metav1.UpdateOptions{})
+		}
+		return err
+	})
+}
+
+func (r *Reconciler) monitoringConfigMap(ctx context.Context) (*v1.ConfigMap, bool, error) {
+	cm, err := r.kubernetescli.CoreV1().ConfigMaps(monitoringName.Namespace).Get(ctx, monitoringName.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      monitoringName.Name,
+				Namespace: monitoringName.Namespace,
+			},
+			Data: map[string]string{
+				"config.yaml": defaultConfig,
+			},
+		}, true, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	return cm, false, nil
+}
+
+func triggerReconcile(cm *corev1.ConfigMap) bool {
+	return cm.Name == monitoringName.Name && cm.Namespace == monitoringName.Namespace
+}
+
+// SetupWithManager setup the manager
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.log.Info("starting starting cluster monitoring controller")
+
+	isMonitoringConfigMap := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			_, ok := e.ObjectOld.(*arov1alpha1.Cluster)
+			if ok {
+				return true
+			}
+
+			cm, ok := e.ObjectOld.(*corev1.ConfigMap)
+			return ok && triggerReconcile(cm)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			_, ok := e.Object.(*arov1alpha1.Cluster)
+			if ok {
+				return true
+			}
+
+			cm, ok := e.Object.(*corev1.ConfigMap)
+			return ok && triggerReconcile(cm)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			_, ok := e.Object.(*arov1alpha1.Cluster)
+			if ok {
+				return true
+			}
+
+			cm, ok := e.Object.(*corev1.ConfigMap)
+			return ok && triggerReconcile(cm)
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}).
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/1173
+		// equivalent to For(&v1.ConfigMap{})., but can't call For multiple times on one builder
+		Watches(&source.Kind{Type: &v1.ConfigMap{}}, &handler.EnqueueRequestForObject{}).
+		Owns(&v1.ConfigMap{}).
```

Comment:
I don't think we need the Owns line.  The pull secret controller has `Owns(&v1.Secret{})` to pick up changes to the Secret associated with the Cluster object - this is n/a in this controller so I think this is just an unnecessary copy/paste.

Reply:
yes, you are right. It  appears working without this line. However I don't quite understand what is the difference between monitoring config map and the other secret (assuming you are referring to operator's reference secret in "openshift-azure-operator" namespace)

Reply:
The pull secret controller code is dependent on the pull secret stored in the operator configuration secret.  Therefore the pull secret controller needs to be aware of changes to the Cluster object, the associated operator configuration secret, and the cluster pull secret itself.

This controller is not dependent on anything stored in the operator configuration secret.  We only need to be aware of changes to the Cluster object and the prometheus configmap.